DBZ-5930: Add the doc for vitess snapshot.mode (#4158)
debezium/debezium: documentation/modules/ROOT/pages/connectors/vitess.adoc

The connector produces a change event for every row-level insert, update, and delete operation that was captured and sends change event records for each table in a separate Kafka topic. Client applications read the Kafka topics that correspond to the database tables of interest, and can react to every row-level event they receive from those topics.

> The underlying MySQL normally purges binary logs after some period of time. This means that the connector does not have the complete history of all changes that have been made to the database. Therefore, when the connector first connects to a particular database, it starts by performing a consistent snapshot of each of the database. After the connector completes the snapshot, it continues streaming changes from the exact point at which the snapshot was made. This way, the connector starts with a consistent view of all of the data, and does not omit any changes that were made while the snapshot was being taken.

Reviewer: ```suggestion
The underlying MySQL implementation in Vitess purges binary logs based on some configurable period of time. 
Because the contents of the binlog might not be complete, the connector requires another mechanism to ensure that it captures the complete content of a particular database.
Thus, when the connector connects to a database for the first time, it performs a consistent snapshot of the database. 
After the connector completes the snapshot, it continues streaming changes from the exact point at which the snapshot was made. 
In this way, the connector starts with a consistent view of all of the data, and does not omit any changes that were made while the snapshot was being taken.
```

> The connector is tolerant of failures. As the connector reads changes and produces events, it records the VGTID position for each event. If the connector stops for any reason (including communication failures, network problems, or crashes), upon restart the connector continues reading the WAL where it last left off. This does not include snapshots. If the connector stops during a snapshot, upon restart the connector does not continue performing snapshots where it last left off.

Reviewer: ```suggestion
The connector is tolerant of failures. 
As the connector reads changes and produces events, it records the VGTID position for each event. 
If the connector stops for any reason (including communication failures, network problems, or crashes), after the connectors restarts, it continues reading the WAL from the point at which it stopped. 
This behavior does not apply to snapshots. If the connector stops during a snapshot, after a restart, the connector does not continue performing snapshots where it last left off.
```

We'll talk later about how the connector behaves xref:{link-vitess-connector}#vitess-when-things-go-wrong[when things go wrong].

// Type: assembly
// ModuleID: how-debezium-vitess-connectors-work
// Title: How {prodname} Vitess connectors work
[[how-the-vitess-connector-works]]
== How the connector works

To optimally configure and run a {prodname} Vitess connector, it is helpful to understand how the connector performs snapshots, streams change events, determines Kafka topic names, and uses metadata.

// Type: concept
// ModuleID: how-debezium-vitess-connectors-perform-snapshot
// Title: How {prodname} Vitess connectors perform a snapshot
[[vitess-snapshot]]
=== Snapshot

> Most MySQL servers are configured to not retain the complete history of the database in the binary logs. This means that the connector would be unable to see the entire history of the database by reading only the binary logs. Consequently, the first time that the connector starts, it performs an initial consistent snapshot of the database. You can change this behavior by setting the xref:{link-vitess-connector}#vitess-property-snapshot-mode[`snapshot.mode` connector configuration property] to a value other than `initial`.

Reviewer: ```suggestion
Typically, a MySQL servers is not configured to retain the complete history of the database in the binary logs. 
As a result, the connector is unable to read the entire history of the database from the binary logs. 
For this reason, the first time that the connector starts, it performs an initial consistent snapshot of the database. 
You can change this behavior by setting the xref:{link-vitess-connector}#vitess-property-snapshot-mode[`snapshot.mode` connector configuration property] to a value other than `initial`.
```

Author: I added [1cc6880](https://github.com/debezium/debezium/pull/4158/commits/1cc688025917d53b931cb5063eb9796761edb3f0).

This snapshot feature is built on link:https://vitess.io/docs/design-docs/vreplication/vstream/vscopy/[VStream Copy] introduced in version 7.0.

[NOTE]
====
In Vitess 15 or earlier, it is not possible to resume `VStream Copy.`

> Starting in Vitess 16, it will support resuming this initial copy phase if it's interrupted for any reason.

Reviewer: ```suggestion
Beginning with Vitess 16, you can resume the initial copy phase if it's interrupted for any reason.
```

It's not clear from the original text, but with Vitess 16 and later, is the initial copy phase resumed automatically following an interruption?

Reviewer 2: @yoheimuta Could you please comment on that question? Thanks a lot.

Author: Thank you for your reviews!

> with Vitess 16 and later, is the initial copy phase resumed automatically following an interruption?

@roldanbob No. In order to achieve this, the connector will also need to be updated.

Reviewer: > Thank you for your reviews!

You're very welcome!

> > with Vitess 16 and later, is the initial copy phase resumed automatically following an interruption?
> 
> @roldanbob No. In order to achieve this, the connector will also need to be updated.

I'm not sure that I follow. Do you mean that:

1. You must restart the connector to resume a snapshot that was interrupted?
2. You must update the Debezium Vitess connector to the version that this change is applied to (2.1?)?
3.  Or is it actually both?: That is, if you're running earlier versions of the connector against Vitess 16 or later, you cannot resume a snapshot that was interrupted. You must be running 2.1 or later _and_ restart the connector to resume an interrupted snapshot.

Author: It means the release of Vitess 16 is a blocker for me to implement this feature fault torelantly.

// Title: How {prodname} Vitess connectors work
[[how-the-vitess-connector-works]]
== How the connector works

To optimally configure and run a {prodname} Vitess connector, it is helpful to understand how the connector performs snapshots, streams change events, determines Kafka topic names, and uses metadata.

// Type: concept
// ModuleID: how-debezium-vitess-connectors-perform-snapshot
// Title: How {prodname} Vitess connectors perform a snapshot
[[vitess-snapshot]]
=== Snapshot

Typically, a MySQL server is not configured to retain the complete history of the database in the binary logs.
As a result, the connector is unable to read the entire history of the database from the binary logs. 
For this reason, the first time that the connector starts, it performs an initial consistent snapshot of the database. 
You can change this behavior by setting the xref:{link-vitess-connector}#vitess-property-snapshot-mode[`snapshot.mode` connector configuration property] to a value other than `initial`.
This snapshot feature is built on link:https://vitess.io/docs/design-docs/vreplication/vstream/vscopy/[VStream Copy] introduced in version 7.0.

[NOTE]
====

> In Vitess 15 or earlier, it is not possible to resume `VStream Copy.`
> Starting in Vitess 16, it will support resuming this initial copy phase if it's interrupted for any reason.

Reviewer: ```suggestion
Beginning with this release of the connector, in environments that run Vitess 16 or later, if a snapshot is interrupted for any reason, after the connection is restored, the snapshot automatically resumes processing data from the point at which it stopped. In environments that run Vitess 15 or earlier, automatic retry is not available. 
```
@yoheimuta Let me know if this latest edit correctly captures the behavior. Per the new description, I assume that the snapshot resumes at the point at which it stopped, yes? That would seem to be the sensible thing. :)

Author: @roldanbob I'm concerned that users may misunderstand that the connector can resume the snapshot just if they update Vitess to 16.
The correct timeline for implementing this feature is as follows

1. Vitess team releases v16.
2. Someone (I or maintainer) updates a connector to build on Vitess 16 feature and cut the release, for example, Debezium v3.
3. Users begins using Vitess v16.
4. Users begins using Debezium v3.

To avoid confusion, it may be best to remove this note. 🙇‍♂️

Author: @roldanbob Thank you for your suggestion!
I added [5da7d4a](https://github.com/debezium/debezium/pull/4158/commits/5da7d4ac2746f8a65848ee0fe32a59bcbbbd19ea).
